sql-plugin/src/main/scala/com/nvidia/spark/rapids/DeviceMemoryEventHandler.scala

```diff
@@ -30,53 +30,129 @@ import org.apache.spark.sql.rapids.execution.TrampolineUtil
+   * synchronizes on a depleted store.
+   */
+  class OOMRetryState {
+    private var synchronizeAttempts = 0
+    private var retryCountLastSynced = 0
+
+    def getRetriesSoFar: Int = synchronizeAttempts
+
+    private def reset(): Unit = {
+      synchronizeAttempts = 0
+      retryCountLastSynced = 0
+    }
+
+    /**
+     * If we have synchronized less times than `maxFailedOOMRetries` we allow
+     * this retry to proceed, and track the `retryCount` provided by cuDF. If we
+     * are above our counter, we reset our state.
+     */
+    def shouldTrySynchronizing(retryCount: Int): Boolean = {
+      if (synchronizeAttempts < maxFailedOOMRetries) {
+        retryCountLastSynced = retryCount
+        synchronizeAttempts += 1
+        true
+      } else {
+        reset()
+        false
+      }
+    }
+
+    /**
+     * We reset our counters if `storeSize` is non-zero (as we only track when the store is
+     * depleted), or `retryCount` is less than or equal to what we had previously
+     * recorded in `shouldTrySynchronizing`. We do this to detect that the new failures
+     * are for a separate allocation (we need to give this new attempt a new set of
+     * retries.)
+     *
+     * For example, if an allocation fails and we deplete the store, `retryCountLastSynced`
+     * is set to the last `retryCount` sent to us by cuDF as we keep allowing retries
+     * from cuDF. If we succeed, cuDF resets `retryCount`, and so the new count sent to us
+     * must be <= than what we saw last, so we can reset our tracking.
+     */
+    def resetIfNeeded(retryCount: Int, storeSize: Long): Unit = {
+      if (storeSize != 0 || retryCount <= retryCountLastSynced) {
+        reset()
+      }
+    }
+  }
+
   /**
    * Handles RMM allocation failures by spilling buffers from device memory.
    * @param allocSize the byte amount that RMM failed to allocate
+   * @param retryCount the number of times this allocation has been retried after failure
    * @return true if allocation should be reattempted or false if it should fail
    */
-  override def onAllocFailure(allocSize: Long): Boolean = {
+  override def onAllocFailure(allocSize: Long, retryCount: Int): Boolean = {
     try {
-      val nvtx = new NvtxRange("onAllocFailure", NvtxColor.RED)
-      try {
+      withResource(new NvtxRange("onAllocFailure", NvtxColor.RED)) { _ =>
         val storeSize = store.currentSize
+        val attemptMsg = if (retryCount > 0) {
+          s"Attempt ${retryCount}. "
+        } else {
+          "First attempt. "
+        }
+
+        val retryState = oomRetryState.get()
+        retryState.resetIfNeeded(retryCount, storeSize)
+
         logInfo(s"Device allocation of $allocSize bytes failed, device store has " +
-            s"$storeSize bytes. Total RMM allocated is ${Rmm.getTotalBytesAllocated} bytes.")
+          s"$storeSize bytes. $attemptMsg" +
+          s"Total RMM allocated is ${Rmm.getTotalBytesAllocated} bytes. ")
         if (storeSize == 0) {
```

Comment:
Not an issue specific to this PR, but I don't think we should be failing/synchronizing only if the store size goes to zero.  `store.currentSize` is the size of all buffers in the store, spillable or not.  If we cannot spill a buffer because it is currently locked/acquired, then I don't think storeSize will go to zero.

Seems like the logic needs to be keyed on "I was not able to spill anything" rather than "store is empty."  We could not spill anything either because there are no buffers to spill (store size == 0) or all remaining buffers in the store are in an unspillable state.

Reply:
In this scenario, `store.currentSize` goes to zero because buffers are getting removed from the `BufferTracker` when we call `synchronousSpill`. So, from the perspective of this PR, it seems we'd reach 0, but there are other bigger issues we want to address and a follow up will be filed to address (e.g. a buffer should not be spilled if it is held). Really we should probably link it with this https://github.com/NVIDIA/spark-rapids/issues/6561 somehow.

Reply:
Yeah turns out store size should always go to zero but for bad reasons.  Filed #6864 to track that.